runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5114,16 +5115,30 @@ class INTERPRETER_CLASS
 			}
 		}
 
+		bp = buildSpecialStackFrame(REGISTER_ARGS, J9SF_FRAME_TYPE_JNI_NATIVE_METHOD, jitStackFrameFlags(REGISTER_ARGS, 0), true);
```

Comment:
You are not calling JNI - this frame should be an INL frame for the current native.

Reply:
As discussed with @tajila (see the comments from https://github.com/eclipse-openj9/openj9/pull/15083#discussion_r902682126), only a JNI native frame works with the situation when a user can call a normal JNI function with `JNIEnv` saved in a global variable which is used in a FFI downcall (verified at https://github.com/eclipse-openj9/openj9/pull/15083#discussion_r919540104).  So an INL frame won't work at this point.

Reply:
I don't believe this is a valid scenario. If you want to use the JNIEnv, use a JNI callout. Storing JNIEnv in global variables is strongly discouraged, since it will almost certainly lead to using the wrong JNIEnv on a thread.

I won't fight using a JNI frame for consistency if a native callin is used.

Reply:
> I don't believe this is a valid scenario. If you want to use the JNIEnv, use a JNI callout. Storing JNIEnv in global variables is strongly discouraged, since it will almost certainly lead to using the wrong JNIEnv on a thread.

While its not recommended to stash JNIEnv pointers, the fact that hotspot supports call-ins in FFI natives means we will need some support for this as well. It might be worth experimenting how hotspots deals with incorrect JNIEnv's when doing call-ins.

Reply:
Using the simple test we used previously for this call-in in FFI native with a tampered JNIEnv (e.g. `globalEnv + 2`) as follows:
[1] JniTest.java
```
import java.lang.invoke.MethodHandle;
...
public class JniTest { 
	private static Linker linker = Linker.nativeLinker();
   static {
      System.loadLibrary("jnitest"); 
   }
	private static final SymbolLookup nativeLibLookup = SymbolLookup.loaderLookup();

   private native int addJNI2Ints(int arg1, int arg2);

	public void test_add2Ints() throws Throwable {
		FunctionDescriptor fd = FunctionDescriptor.of(JAVA_INT, JAVA_INT, JAVA_INT);
		Addressable functionSymbol = nativeLibLookup.lookup("add2Ints").get();
		MethodHandle mh = linker.downcallHandle(functionSymbol, fd);
		int result = (int)mh.invokeExact(112, 123);
		System.out.println("test_add2Ints result = " + result);
	}

   public static void main(String[] args) throws Throwable {
      JniTest test = new JniTest();
      int sum = test.addJNI2Ints(1, 2); 
      System.out.println("addJNI2Ints sum = " + sum);
      test.test_add2Ints();
   }
}
```
[2] JniTest.c
```
#include <jni.h>
#include "JniTest.h"

JNIEnv *globalEnv;
jobject globalObj;

JNIEXPORT int JNICALL Java_JniTest_addJNI2Ints(JNIEnv *env, jobject thisObj, jint arg1, jint arg2) {
    jint intSum = arg1 + arg2;
    globalEnv = env;
    globalObj = thisObj;
    printf("\nJava_JniTest_addJNI2Ints: env = %p, thisObj = %p\n", globalEnv, globalObj);
    return intSum;
}

int
add2Ints(int intArg1, int intArg2)
{
	int intSum = intArg1 + intArg2;
        printf("add2Ints: intSum = %d", intSum);
	printf("\nadd2Ints: env = %p, thisObj = %p\n", globalEnv, globalObj);
	jobject result;
	for (int i = 0; i < 20; i++) {
	  result = (*globalEnv)->NewLocalRef(globalEnv + 2, <------- passed in an incorrect JNIEnv value
                                                                    globalObj);
	  printf("\ni = %d\n", i);
	}
	return intSum;
}
```
Hotspot just crashed in this case
```
../jdk19_hotspot_openjdk_x64/bin/java  --enable-preview  -Djava.library.path=.:  --enable-native-access=ALL-UNNAMED -Dforeign.restricted=permit   JniTest

Java_JniTest_addJNI2Ints: env = 0x7fef040260b0, thisObj = 0x7fef09cea9a8
addJNI2Ints sum = 3
add2Ints: intSum = 235
add2Ints: env = 0x7fef040260b0, thisObj = 0x7fef09cea9a8
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fef0ab04b4d, pid=33217, tid=33218
#
... Aborted                 (core dumped)
```
which indicates that there is no special handling but generated core dump on Hotspot when doing call-ins with a corrupted `JNIEnv` in FFI natives.

Reply:
The JNI spec essentially says that if you do something invalid, you'll crash. Note that -XX:DebugVMAccess on the OpenJ9 VM will detect many of these errors.

Reply:
I tried `-Xcheck:jni` on Hotspot which captured the error as invalid as follows prior to crash:
```
../jdk19_hotspot_openjdk/bin/java -Xcheck:jni --enable-preview -Djava.library.path=.: --enable-native-access=ALL-UNNAMED -Dforeign.restricted=permit JniTest

ava_JniTest_addJNI2Ints: env = 0x7f41d80260e0, thisObj = 0x7f41deade978
addJNI2Ints sum = 3
add2Ints: intSum = 235
add2Ints: env = 0x7f41d80260e0, thisObj = 0x7f41deade978
FATAL ERROR in native method: Using JNIEnv in the wrong thread <-------------
	at java.lang.invoke.LambdaForm$MH/0x0000000801011400.invoke(java.base@19-ea/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x0000000801015c00.invokeExact_MT(java.base@19-ea/LambdaForm$MH)
	at jdk.internal.foreign.abi.DowncallStub/0x0000000801011800.invoke(java.base@19-ea/Unknown Source)
	at java.lang.invoke.LambdaForm$DMH/0x0000000801012000.invokeStaticInit(java.base@19-ea/LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/0x0000000801014c00.invoke(java.base@19-ea/LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/0x0000000801015000.invokeExact_MT(java.base@19-ea/LambdaForm$MH)
	at JniTest.test_add2Ints(JniTest.java:26)
	at JniTest.main(JniTest.java:34)
./test_jni_jdk19.sh: line 7: 35895 Aborted                 (core dumped)
```
But there was no error message for reminder on our build with both `-XX:+DebugVMAccess` and `-Xcheck:jni`
[1] -XX:+DebugVMAccess
```
../jdk19_openj9/bin/java -XX:+DebugVMAccess  --enable-preview  -Djava.library.path=.:  --enable-native-access=ALL-UNNAMED -Dforeign.restricted=permit   JniTest

Java_JniTest_addJNI2Ints: env = 0x98200, thisObj = 0x18e2c8
addJNI2Ints sum = 3
add2Ints: intSum = 235
add2Ints: env = 0x98200, thisObj = 0x18e2c8
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
Handler1=0000759E3D8B9238 Handler2=0000759E3D670C00
R0=0000759E3D8C3FE8 R1=0000759E3DB7CAA0 R2=0000759E3DA46E00 R3=0000000000098210
...
FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
Module=/root/hostdir/jdk19_openj9_ffi_uma_ppc64le_v09_8_3/lib/compressedrefs/libj9vm29.so
Module_base_address=0000759E3D770000
Target=2_90_20220816_000000 (Linux 4.15.0-191-generic)
CPU=ppc64le (8 logical CPUs) (0x3fd6c0000 RAM)
----------- Stack Backtrace -----------
...
mainSynchSignalHandler+0x294 (0x0000759E3D670E94 [libj9prt29.so+0x40e94])
__kernel_sigtramp_rt64+0x0 (0x0000759E3DFA04D8)
 (0x0000000000000000 [<unknown>+0x0])
newLocalRef+0x48 (0x0000759E3D8C3FE8 [libj9vm29.so+0x153fe8]) <--------
add2Ints+0xdc (0x0000759E12B308E0 [libjnitest.so+0x8e0])
 (0x0000759E3D9E91D4 [libj9vm29.so+0x2791d4])
ffi_call_int+0x108 (0x0000759E3D9E6F24 [libj9vm29.so+0x276f24])
ffi_call+0x44 (0x0000759E3D9E6FF8 [libj9vm29.so+0x276ff8])
_ZN32VM_BytecodeInterpreterCompressed38inlInternalDowncallHandlerInvokeNativeERPmRPh+0x55c (0x0000759E3D79F3D0 [libj9vm29.so+0x2f3d0])
_ZN32VM_BytecodeInterpreterCompressed3runEP10J9VMThread+0xfc1c (0x0000759E3D7BBF30 [libj9vm29.so+0x4bf30])
bytecodeLoopCompressed+0x44 (0x0000759E3D781788 [libj9vm29.so+0x11788])
 (0x0000759E3D8FC028 [libj9vm29.so+0x18c028])
runCallInMethod+0x3b8 (0x0000759E3D8929F4 [libj9vm29.so+0x1229f4])
gpProtectedRunCallInMethod+0x78 (0x0000759E3D8C1870 [libj9vm29.so+0x151870])
...
---------------------------------------
```
[2] -Xcheck:jni
```
../jdk19_openj9/bin/java -Xcheck:jni  --enable-preview  -Djava.library.path=.:  --enable-native-access=ALL-UNNAMED -Dforeign.restricted=permit   JniTest
JVMJNCK001I JNI check utility installed. Use -Xcheck:jni:help for usage

Java_JniTest_addJNI2Ints: env = 0x98200, thisObj = 0x18e2c8
addJNI2Ints sum = 3
add2Ints: intSum = 235
add2Ints: env = 0x98200, thisObj = 0x18e2c8
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
Handler1=00007F104F039238 Handler2=00007F104EDF0C00
R0=00007F104D601300 R1=00007F104F2FC870 R2=00007F104D647E00 R3=00007F104D61F0C0
...
FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
Module=/root/hostdir/jdk19_openj9_ffi_uma_ppc64le_v09_8_3/lib/compressedrefs/libj9jnichk29.so
Module_base_address=00007F104D5F0000
Target=2_90_20220816_000000 (Linux 4.15.0-191-generic)
CPU=ppc64le (8 logical CPUs) (0x3fd6c0000 RAM)
----------- Stack Backtrace -----------
protectedBacktrace+0x4c (0x00007F104EE2CDA8 [libj9prt29.so+0x7cda8])
....
mainSynchSignalHandler+0x294 (0x00007F104EDF0E94 [libj9prt29.so+0x40e94])
__kernel_sigtramp_rt64+0x0 (0x00007F104F7204D8)
 (0x0000000000000000 [<unknown>+0x0])
checkNewLocalRef+0x5c (0x00007F104D601300 [libj9jnichk29.so+0x11300]) <-------
add2Ints+0xdc (0x00007F0FFBEF08E0 [libjnitest.so+0x8e0])
 (0x00007F104F1691D4 [libj9vm29.so+0x2791d4])
ffi_call_int+0x108 (0x00007F104F166F24 [libj9vm29.so+0x276f24])
ffi_call+0x44 (0x00007F104F166FF8 [libj9vm29.so+0x276ff8])
_ZN32VM_BytecodeInterpreterCompressed38inlInternalDowncallHandlerInvokeNativeERPmRPh+0x55c (0x00007F104EF1F3D0 [libj9vm29.so+0x2f3d0])
_ZN32VM_BytecodeInterpreterCompressed3runEP10J9VMThread+0xfc1c (0x00007F104EF3BF30 [libj9vm29.so+0x4bf30])
bytecodeLoopCompressed+0x44 (0x00007F104EF01788 [libj9vm29.so+0x11788])
 (0x00007F104F07C028 [libj9vm29.so+0x18c028])
runCallInMethod+0x3b8 (0x00007F104F0129F4 [libj9vm29.so+0x1229f4])
gpProtectedRunCallInMethod+0x78 (0x00007F104F041870 [libj9vm29.so+0x151870])
signalProtectAndRunGlue+0x50 (0x00007F104F0C41AC [libj9vm29.so+0x1d41ac])
...
```
My understanding is that this error should be captured at least in `checkNewLocalRef()` (similar to Hotspot).